core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt

```diff
@@ -0,0 +1,756 @@
+                    allowedTypes.addAll(allLoadingGaugeTypeSet)
+                } else {
+                    allowedTypes.addAll(compatibleTypes)
+                }
+            }
+        }
+        res.put(begin.meters, end.meters, fromAllowedSet(allowedTypes))
+    }
+    return res
+}
+
+private fun buildCompatibleGaugeTypesMap(): Map<RJSLoadingGaugeType, Set<RJSLoadingGaugeType>> {
+    val compatibleGaugeType = hashMapOf<RJSLoadingGaugeType, Set<RJSLoadingGaugeType>>()
+    for (gaugeType in enumValues<RJSLoadingGaugeType>()) {
+        val compatibleTypes = gaugeType.compatibleGaugeTypes
+        if (compatibleTypes != null) compatibleGaugeType[gaugeType] = compatibleTypes
+    }
+    return compatibleGaugeType
+}
+
+val compatibleGaugeTypeMap = buildCompatibleGaugeTypesMap()
+val allLoadingGaugeTypeSet = enumValues<RJSLoadingGaugeType>().toSet()
+
+private fun parseRjsRouteWaypoint(
+    rjsDetector: RJSRouteWaypoint,
+    trackSectionNameToDistanceSortedDetectors:
+        MutableMap<String, TreeMap<Offset<TrackSection>, MutableList<String>>>,
+) {
+    val trackDetectors =
+        trackSectionNameToDistanceSortedDetectors.computeIfAbsent(rjsDetector.track) { TreeMap() }
+    val signalPosition = Offset<TrackSection>(rjsDetector.position.meters)
+    val detectorNames = trackDetectors.computeIfAbsent(signalPosition) { mutableListOf() }
+    detectorNames.add(rjsDetector.id)
+}
+
+/** Build detection zones on a partially-filled builder (track-sections, nodes and detectors) */
+private fun buildZones(builder: RawInfraFromRjsBuilder) {
+    // run a union-find to associate track endpoints:
+    // union by node
+    val trackNodePool = builder.getTrackNodes()
+    val uf = UnionFind(builder.getTrackSections().size.toInt() * 2)
+    for (nodeIdx in trackNodePool) {
+        val ports = builder.getTrackNodePorts(nodeIdx)
+        var firstEndpoint: EndpointTrackSectionId? = null
+        for (port in ports) {
+            val endpoint = ports[port]
+            if (firstEndpoint == null) {
+                firstEndpoint = endpoint
+            } else {
+                uf.union(firstEndpoint.index.toInt(), endpoint.index.toInt())
+            }
+        }
+    }
+    // union by track-section without detector
+    for (trackSectionIdx in builder.getTrackSections()) {
+        val detectors = builder.getTrackSectionDetectors(trackSectionIdx)
+        if (detectors.size == 0) {
+            val startEndpoint = EndpointTrackSectionId(trackSectionIdx, Endpoint.START)
+            val endEndpoint = EndpointTrackSectionId(trackSectionIdx, Endpoint.END)
+            uf.union(startEndpoint.index.toInt(), endEndpoint.index.toInt())
+        }
+    }
+
+    // now that union-find is complete, create detection zones
+    val rootToZoneMap = mutableMapOf<Int, ZoneId>()
+    for (trackSectionIdx in builder.getTrackSections()) {
+        val detectors = builder.getTrackSectionDetectors(trackSectionIdx)
+
+        // prepare "node" detection zones
```

Comment:
What are those?

Reply:
the algorithm is correct but weird, I'll refactor it